### DISCUSSION TITLE
How to update and store Provider access token and refresh token?

### DISCUSSION THREAD

**Discussion by ggogobera** [supabase/supabase#22578] [Questions] [answered]
Supabase doesn't seem to automatically refresh the provider access token.

I'm using Zoom as an OAuth provider. However, after 1 hour, both the provider's `access_token` and `refresh_token` become `undefined`. I still need to figure out why they vanish from the session object after this duration which is its own issue.

To implement a custom refresh mechanism, how can I update the session with new tokens?

It appears that the `setSession` method only takes the `access_token` and `refresh_token` specific to GoTrue, not those from the provider.

**Comment by GaryAustin1** [comment, marked as answer]
Supabase does not store them for security reasons.
https://supabase.com/docs/guides/auth/social-login#provider-tokens
![image](https://github.com/supabase/supabase/assets/54564956/cc2e2dec-4cd2-48b7-85c7-1d2ea27412f8)


You can store them however you like in local storage, etc.

**Reply by GaryAustin1** [reply]
https://supabase.com/docs/reference/javascript/auth-signinwithoauth?example=sign-in-with-scopes

This looks like an example of storing them.

**Reply by ggogobera** [reply]
Makes total sense. I should have read the docs more carefully!

**Comment by MaestroDev19** [comment]
what if your auth is server side , im using nextjs

**Reply by ladparth** [reply]
Were you able to figure this out?

**Reply by markdawson** [reply]
I was able to get the provider token in my callback endpoint like this. This is a slightly modified code snippet from [this tutorial](https://supabase.com/docs/guides/auth/social-login/auth-github?queryGroups=environment&environment=server&queryGroups=framework&framework=nextjs).

```ts
// auth/callback/route.ts

import { createClient } from "@/utils/supabase/server";
import { NextResponse } from "next/server";

export async function GET(request: Request) {
  const { searchParams, origin } = new URL(request.url)
  const code = searchParams.get('code')
  // if "next" is in param, use it as the redirect URL
  const next = searchParams.get('next') ?? '/'

  if (code) {
    console.log("Got code", code)

    const supabase = await createClient()
    // ❗❗❗ Modify exchangeCodeForSession to get data
    const { error, data } = await supabase.auth.exchangeCodeForSession(code)
    console.log("Provider token", data.session?.provider_token)
    // ❗❗❗ Insert logic here to save provider token somewhere.
    

    if (!error) {
      const forwardedHost = request.headers.get('x-forwarded-host') // original origin before load balancer
      const isLocalEnv = process.env.NODE_ENV === 'development'
      if (isLocalEnv) {
        // we can be sure that there is no load balancer in between, so no need to watch for X-Forwarded-Host
        return NextResponse.redirect(`${origin}${next}`)
      } else if (forwardedHost) {
        return NextResponse.redirect(`https://${forwardedHost}${next}`)
      } else {
        return NextResponse.redirect(`${origin}${next}`)
      }
    }
  }

  // return the user to an error page with instructions
  return NextResponse.redirect(`${origin}/auth/auth-code-error`)
}
```

**Comment by zeluspudding** [comment]
Feels strange that Supabase doesn't default to use Vault to store access tokens for us. Like why did you make Vault if not to make this possible for us? Am I wrong that they seem to be stepping away from promoting this? It used to be you could create encrypted fields/columns in a table but they removed that... and it's somewhat tricky/annoying to setup oneself. 

I'm assuming _that is_ the correct way to go about this... can someone correct me if I'm wrong.

**Reply by BugAnnihilator** [reply]
correct!
